timemachine/md/hrex.py

```diff
@@ -238,6 +239,26 @@ def state_replica_pairs(self) -> List[Tuple[StateIdx, Replica]]:
         return [(StateIdx(i), self.replicas[replica_idx]) for i, replica_idx in enumerate(self.replica_idx_by_state)]
 
 
+def get_normalized_kl_divergence(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> float:
+    """Heuristic for the how uniformly windows were sampled within an HREX simulation.
+    Based on Eq. 5 from [1], but after summing up the divergences of each state, take the mean
+    so that it is possible to compare the values between simulations of different numbers of windows.
+
+    A value closer to 0.0 indicates more uniforming sampling.
+
+    References
+    ----------
+    [1]: https://doi.org/10.1021/acs.jctc.0c00660, https://chemrxiv.org/engage/chemrxiv/article-details/60c74d2e702a9b007018b7ef
+    """
+    cumulative_counts = get_cumulative_replica_state_counts(replica_idx_by_state_by_iter)
+    n_iters, n_states, n_replicas = cumulative_counts.shape
+    count_by_replica_by_state = cumulative_counts[-1]
+    fraction_by_replica_by_state = count_by_replica_by_state / n_iters
+    uniform_dist = np.ones_like(fraction_by_replica_by_state) / n_states
+    # No need to add eps to the fraction_by_replica_state, as kel_entr returns a KL of 0 in that case
+    return np.sum(rel_entr(fraction_by_replica_by_state, uniform_dist)) / n_states
```

Comment:
A few notes I had when trying to interpret this; unsure if they would help clarify the implementation:

1. It made sense to me to think of this as averaging over replicas the KL divergence between the state distribution for a given replica and the uniform distribution (rather than the KL divergence between 2-d distributions). I.e. thinking of it as the (equivalent)

    ```python
    np.mean(np.sum(rel_entr(fraction_by_replica_by_state, uniform_dist), axis=1))
    ```

3. This is also seems equivalent to

    ```python
    np.mean(entropy(fraction_by_replica_by_state, uniform_dist, axis=1))
    ```

5. Finally, I think you can avoid constructing the uniform distribution and just write this as

    ```python
    -np.mean(entropy(fraction_by_replica_by_state, axis=1)) + np.log(n_states)
    ```

    (using the identity $\sum_i p_i \log (p_i / u_i) =  \sum_i p_i \log p_i + \log N = -H_p + \log N$, where $H_p$ is the entropy of $p$)

Reply:
Nice! The last variant is great, added it in https://github.com/proteneer/timemachine/pull/1314/commits/43a6b9a54f7bee3dc09477ef6a0bb5fe7edd0fd7 and added a note about the construction.

Reply:
One last option you might prefer:

```python
-entropy(fraction_by_replica_by_state, axis=None) + 2 * np.log(n_states)
```

(`scipy.stats.entropy` has a sneaky default of `axis=0`, which seems different from typical the numpy convention; might prefer to use the above or explicitly pass `axis=0` if it looks confusing)

Reply:
> The `2 * np.log(n_states)` isn't immediately obvious to me, so I added `axis=0` to be explicit in [05adc51](https://github.com/proteneer/timemachine/commit/05adc5102481746ae26febacf84a4c79664471a5)

I think this is equally clear. (The extra factor of 2 comes from `N` being `n_states ** 2` in the latter case, since we're viewing the distribution as  2-d)